app/services/catalog/update_order_item.rb

```diff
@@ -0,0 +1,43 @@
+module Catalog
+  class UpdateOrderItem
+    class OrderItemNotFound < StandardError; end
+
+    def initialize(topic)
+      @payload = topic.payload
+      @message = topic.message
+    end
+
+    def process
+      Rails.logger.info("Processing service order topic message: #{@message} with payload: #{@payload}")
+
+      Rails.logger.info("Searching for OrderItem with a task_ref: #{@payload["task_id"]}")
+      order_item = find_order_item
+      Rails.logger.info("Found OrderItem: #{order_item.id}")
+
+      ManageIQ::API::Common::Request.with_request(order_item.context.transform_keys(&:to_sym)) do
+        order_item.update_message("info", "Task update message received with payload: #{@payload}")
+
+        if @payload["state"] == "completed"
+          order_item.state = "Order Completed"
+          order_item.update_message("info", "Order Complete")
+
+          Rails.logger.info("Updating OrderItem: #{order_item.id} with 'Order Completed' state")
+          order_item.save!
+          Rails.logger.info("Finished updating OrderItem: #{order_item.id} with 'Order Completed' state")
+        end
+      end
+    end
+
+    private
+
+    def find_order_item
+      order_item = OrderItem.where(:topology_task_ref => @payload["task_id"]).first
```

Comment:
@eclarizio Can a tap be used here, also couldn't this just be a find! and we catch the ActiveRecord::RecordNotFound do we need a special exception, because the function name itself is descriptive

```ruby
def find_order_item
  OrderItem.find_by!(topology_task_ref => @payload["task_id"])
rescue ActiveRecord::RecordNotFound
  Rails.logger.error("Could not find an OrderItem with topology_task_ref: #{@payload["task_id"]}")
  raise "Could not find an OrderItem with topology_task_ref: #{@payload["task_id"]}"
end
```

Reply:
Yeah, I changed this to work this way in #221.